# Kindel – indel-aware offline consensus calling for DNA alignments
[![PyPI version](https://badge.fury.io/py/kindel.svg)](https://badge.fury.io/py/kindel)

Kindel is a simple consensus caller which takes a SAM/BAM file as input and generates a *single chromosome* consensus sequence in fasta format. Unlike similar tools it correctly reconciles small indels described in CIGAR fields. **An experimental feature `--fix-gaps` can also reconcile longer gaps and consensus indels by leveraging context from unaligned (soft-clipped) regions of sequences.**

Existing consensus calling approaches are complicated and often involve a variant calling step. While an [elegant and sophisticated streaming approach](https://github.com/karel-brinda/ococo) was recently released, it cannot reconcile indels.


## Installation
```
pip3 install kindel
```
Dependencies should automatically installed, except for Samtools which is needed for BAM input.  
Otherwise install using a tagged release as master may well be broken.


## Usage
### Command line
```
$ kindel --fix-gaps --min-depth 10 alignment.bam
```
The consensus fasta is sent to `stdout` and a report to `stderr`
```
$ kindel -h
usage: kindel.py [-h] [-f] [--trim-ends] [--threshold-weight THRESHOLD_WEIGHT]
                 [-m MIN_DEPTH] [-c CLOSURE_K]
                 bam-path

positional arguments:
  bam-path              path to SAM/BAM file

optional arguments:
  -h, --help            show this help message and exit
  -f, --fix-gaps        attempt to reconcile reference at soft-clip boundaries
                        (default: False)
  --trim-ends           trim ambiguous nucleotides (Ns) from sequence ends
                        (default: False)
  --threshold-weight THRESHOLD_WEIGHT
                        consensus threshold weight (default: 0.5)
  -m MIN_DEPTH, --min-depth MIN_DEPTH
                        substitute Ns at coverage depths beneath this value
                        (default: 2)
  -c CLOSURE_K, --closure-k CLOSURE_K
                        match length required to close soft-clipped gaps
                        (default: 7)
```

### Python3
```
from kindel import kindel

kindel.bam_to_consensus_seqrecord(bam_path, threshold_weight=0.5, min_depth=10) # returns BioPython SeqRecord
kindel.bam_to_consensus_fasta(bam_path, threshold_weight=0.5, min_depth=10) # returns fasta string
```

## Issues
Please let me know if you run into problems by opening a GitHub issue, [tweeting](https://twitter.com/beconstant) or mailing me via `b at bede dawt im`.
- Master is ~~probably~~ almost certainly broken – use PYPI or a tagged release
- Conceived for use with viral genomes – requires a single chromosome reference sequence
- Tested only with alignments generated by BWA
- SAM/BAM files must contain an SQ header line containing the reference sequence length.
- In current form `--fix-gaps` can require multiple runs to converge on the optimal consensus
- Slow (10-20k records/s)

## Features
- Consensus gap closure using soft-clipped alignment context ✓* *work in progress*
  - left-to-right gap closure ✓ (experimental; not in tagged release)
  - right-to-left gap closure ✓
  - Meet-in-the-middle gap closure ✓
