# Kindel – indel-aware offline consensus calling for nucleotide alignments
[![PyPI version](https://badge.fury.io/py/kindel.svg)](https://badge.fury.io/py/kindel)

Kindel is a simple consensus caller which takes a SAM/BAM file as input and generates a *single chromosome* consensus sequence in fasta format. Unlike similar tools it correctly reconciles small indels described in CIGAR fields. **An experimental feature `--fix-gaps` can also reconcile longer gaps and consensus indels by leveraging context from unaligned (soft-clipped) regions of sequences.**

Existing consensus calling approaches are complicated and often involve a variant calling step. While an [elegant and sophisticated streaming approach](https://github.com/karel-brinda/ococo) was recently released, it cannot reconcile indels.


## Installation
```
pip3 install kindel
```
Dependencies should automatically installed, except for Samtools which is needed for BAM input.


## Usage
### Command line
```
$ kindel consensus --fix-gaps --min-depth 10 alignment.bam
```
The consensus fasta is sent to `stdout` and a report to `stderr`
```
$ kindel -h
usage: kindel [-h] {consensus,weights,variants} ...

positional arguments:
  {consensus,weights,variants}
    consensus           Infer consensus sequence(s) from alignment in SAM/BAM
                        format
    weights             Returns DataFrame of per-site nucleotide frequencies
                        and coverage
    variants            Output variants exceeding specified absolute and
                        relative frequency thresholds

optional arguments:
  -h, --help            show this help message and exit
```
---
```
$ kindel consensus -h
usage: kindel consensus [-h] [-f] [--trim-ends] [--threshold THRESHOLD]
                        [-m MIN_DEPTH] [-c CLOSURE_K] [-u]
                        bam-path

Infer consensus sequence(s) from alignment in SAM/BAM format

positional arguments:
  bam-path              path to SAM/BAM file

optional arguments:
  -h, --help            show this help message and exit
  -f, --fix-gaps        attempt to reconcile reference at soft-clip boundaries
                        (default: False)
  --trim-ends           trim ambiguous nucleotides (Ns) from sequence ends
                        (default: False)
  --threshold THRESHOLD
                        consensus threshold weight (default: 0.5)
  -m MIN_DEPTH, --min-depth MIN_DEPTH
                        substitute Ns at coverage depths beneath this value
                        (default: 2)
  -c CLOSURE_K, --closure-k CLOSURE_K
                        match length required to close soft-clipped gaps
                        (default: 7)
  -u, --uppercase       close gaps using uppercase alphabet (default: False)
```
---
```
$ kindel weights -h
usage: kindel weights [-h] [-r] [-n] bam-path

Returns DataFrame of per-site nucleotide frequencies and coverage

positional arguments:
  bam-path             path to SAM/BAM file

optional arguments:
  -h, --help           show this help message and exit
  -r, --relative       output relative nucleotide frequencies (default: False)
  -n, --no-confidence  skip confidence calculation (default: False)

```
---
```
$ kindel variants -h
usage: kindel variants [-h] [-a ABS_THRESHOLD] [-r REL_THRESHOLD] [-o]
                       bam-path

Output variants exceeding specified absolute and relative frequency thresholds

positional arguments:
  bam-path              path to SAM/BAM file

optional arguments:
  -h, --help            show this help message and exit
  -a ABS_THRESHOLD, --abs-threshold ABS_THRESHOLD
                        absolute frequency (0-∞) threshold above which to call
                        variants (default: 1)
  -r REL_THRESHOLD, --rel-threshold REL_THRESHOLD
                        relative frequency (0.0-1.0) threshold above which to
                        call variants (default: 0.01)
  -o, --only-variants   exclude invariant sites from output (default: False)
```


### Python3
```python
from kindel import kindel

kindel.bam_to_consensus(bam_path, threshold_weight=0.5, min_depth=10) # returns BioPython SeqRecord
```

## Issues
Please let me know if you run into problems by opening a GitHub issue, [tweeting](https://twitter.com/beconstant) or mailing me via `b at bede dawt im`.
- Master may be broken – use PYPI or a tagged release
- Conceived for use with viral genomes – requires a single chromosome reference sequence
- Tested only with alignments generated by BWA
- SAM/BAM files must contain an SQ header line containing the reference sequence length.
- `--fix-gaps` may require multiple runs to converge on the optimal consensus
- Slow (10-20k records/s)

## Features
- [x] Reconciliation of CIGAR described insertions and deletions
- [x] Gap closure (`--fix-gaps`) using soft-clipped alignment context (not yet in tagged release)
- [ ] Support / test against SAMs from a variety of aligners
- [ ] Support mutiple reference sequences (needs testing)
- [ ] Substitution only mode (`--ignore-indels`)
- [ ] Frequency based variant calling with `kindel variants`
- [ ] Plotting